src/Network/HTTP/Affjax.purs

```diff
@@ -0,0 +1,158 @@
+import Data.Either (Either(..))
+import Data.Foreign (Foreign(..), F())
+import Data.Function (Fn4(), runFn4)
+import Data.Options (Option(), Options(), options, (:=), opt)
+import Network.HTTP.Affjax.Request
+import Network.HTTP.Affjax.Response
+import Network.HTTP.Affjax.ResponseType
+import Network.HTTP.Method (Method(..))
+import Network.HTTP.RequestHeader (RequestHeader())
+import Network.HTTP.ResponseHeader (ResponseHeader(), responseHeader)
+import Network.HTTP.StatusCode (StatusCode())
+
+-- | The effect type for AJAX requests made with Affjax.
+foreign import data Ajax :: !
+
+-- | The type for Affjax requests.
+type Affjax e a = Aff (ajax :: Ajax | e) (AffjaxResponse a)
+
+-- | Options type for Affjax requests.
+foreign import data AffjaxOptions :: *
+
+-- | The type of records that will be received as an Affjax response.
+type AffjaxResponse a =
+  { status :: StatusCode
+  , headers :: [ResponseHeader]
+  , response :: a
+  }
+
+-- | Sets the URL for a request.
+url :: Option AffjaxOptions String
+url = opt "url"
+
+-- | Sets the HTTP method for a request.
+method :: Option AffjaxOptions Method
+method = opt "method"
+
+-- | Sets the content to send in a request.
+content :: Option AffjaxOptions RequestContent
+content = opt "content"
+
+-- | Sets the headers to send with a request.
+headers :: Option AffjaxOptions [RequestHeader]
+headers = opt "headers"
+
+-- | Sets the HTTP auth username to send with a request.
+username :: Option AffjaxOptions String
+username = opt "username"
+
+-- | Sets the HTTP auth password to send with a request.
+password :: Option AffjaxOptions String
+password = opt "password"
+
+-- | Sets the expected response type for a request. This is not exposed outside
+-- | of the module as the `ResponseType` is set based on the `Responsable`
+-- | instance for the expected result content type.
+responseType = opt "responseType" :: Option AffjaxOptions ResponseType
+
+-- | Runs a request.
+affjax :: forall e a. Responsable a ->
+                      Options AffjaxOptions ->
+                      Affjax e a
+affjax r = makeAff <<< affjax' r
+
+-- | Runs a request directly in Eff.
+affjax' :: forall e a. Responsable a ->
+                       Options AffjaxOptions ->
+                       (Error -> Eff (ajax :: Ajax | e) Unit) ->
+                       (AffjaxResponse a -> Eff (ajax :: Ajax | e) Unit) ->
+                       Eff (ajax :: Ajax | e) Unit
+affjax' (Responsable read ty) opts eb cb =
+  runFn4 unsafeAjax responseHeader (options $ opts <> responseType := ty) eb cb'
+  where
+  cb' :: AffjaxResponse Foreign -> Eff (ajax :: Ajax | e) Unit
+  cb' res = case res { response = _  } <$> read res.response of
+    Left err -> eb $ error (show err)
+    Right res' -> cb res'
+
+get :: forall e a. Responsable a -> String -> Affjax e a
+get r addr = affjax r $ method := GET
+                     <> url := addr
+
+post :: forall e a. Responsable a -> String -> RequestContent -> Affjax e a
```

Comment:
> The issue I had is it would give us orphan instances for Blob, FormData, Document and all the ArrayView types - they need IsOption instances.

This seems like a limitation of the options library (or our use of the library). Also I'd note the only value this library is providing us with is a monoidal syntax for combining optional options. Which is of questionable benefit given all the FFI magic going on. I'd almost prefer the "base" version just use a record with `Maybe`s for everything, that way you could provide a default with `Nothing` for everything and users could use update syntax to make changes they want. Then other modules could build on that simple dumb version and do things like free monad DSL, options-style monoidal syntax sugar, etc.

I'd just *really* like the signature to be something like:

```purescript
affjax :: forall e a b. (Requestable a, Respondable b) => AjaxRequest a -> Aff (ajax :: AJAX | e) (AjaxResponse b)
```

with `affjax` function handling setting `Content-Type` and `Accept` headers based on the given requestable / respondable (of course, these could be overridden by the user if necessary).

By the way, to some extent I recognize this as bikeshedding, so happy to merge if you like the API as it is. :smile: